### ISSUE TITLE
Synchronize Playback with BPM (Bar/Beats)

### ISSUE THREAD

**Issue by reduz** [godotengine/godot-proposals#8937]
### Describe the project you are working on

Godot

### Describe the problem or limitation you are having in your project

Users often request the desire to synchronize stream playback (triggering) with BPM, which is often desirable for rythm games. Currently Godot does not offer any reliable way to do it.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

A new "bpm_sync" property should be added to AudioStreamPlayer/2D/3D that lets you customize BPM synchronization.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

This is a new "bpm_sync" property to be added to AudioStreamPlayer/2D/3D, it contains the follow values in an enum:

* **Sync Source**: This stream is the sync source. Other streams will synchronize to it.
* **Sync to Beat**: This stream will synchronize to play at the earlier next beat from _Sync Source_.
* **Sync to Bar**: This stream will synchronize to play at the earlier next bar from _Sync Source_.

And that's it.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

N/A

### Is there a reason why this should be core and not an add-on in the asset library?

N/A

**Comment by Mickeon** [comment]
While I do not necessarily disagree, these 3 classes are getting quite "bloated" with features.
It wouldn't be so much of an issue if, for simpler use cases, you could access the **AudioServer** and instantiate your own playbacks directly. This is not currently possible, so for every use-case, you always have to instantiate an **AudioStreamPlayer** node in the scene tree. One that is getting increasingly more packed with features.

**Comment by reduz** [comment]
@Mickeon I disagree with that take. I think AudioStreamPlayer is relatively small (and now @KoBeWi is unifying the 3 classes), and on the Godot side the vast majority of work and features is pretty well distributed in AudioStreams.

Keep in mind this feature is just a simple enum on the side of AudioStreamPlayer, the core of the work is inside AudioServer.

**Comment by NoodleSushi** [comment]
Is this upcoming feature going to take audio that changes BPM at specific points of time into account? Or will this require a custom implementation?

**Comment by midiphony** [comment]
This might go off-topic, but if it is possible to synchronize an audio stream to the next beat or bar of another, would it be possible to synchronize a stream more arbitrarily ?

For example, Unity exposes an API to schedule the start of an audio source on the audio thread timeline, driven by the DSP clock.
https://docs.unity3d.com/ScriptReference/AudioSource.PlayScheduled.html

This allows to stitch together audio clips accurately. The use case is roughly summarized in this presentation by Jakob Schmid, audio composer/programmer of the game "140" (starting from slide 35) : 
https://schmid.dk/talks/2014-05-23-Nordic_Game_Conference/Schmid-2014-05-23-NGC-140.pdf

This would allow accurate audio synchronization without being restricted to specific use cases

**Comment by reduz** [comment]
> Is this proposal going to take audio that changes BPM at specific points of time into account? Or will that require a custom implementation?

Don't see why it should not work.

**Comment by reduz** [comment]
@midiphony this kind of sucks honestly and I think ita better handled by the interactive audio PR. You cant simply stich audio, you have to fade it somehow. Its not as simple.

**Comment by reduz** [comment]
That said, enqueueing to a position of another stream should be definitely possible.

**Comment by KoBeWi** [comment]
> and now @KoBeWi is unifying the 3 classes

It was already merged: https://github.com/godotengine/godot/pull/87061

**Comment by NoodleSushi** [comment]
> This might go off-topic, but if it is possible to synchronize an audio stream to the next beat or bar of another, would it be possible to synchronize a stream more arbitrarily ?
> 
> For example, Unity exposes an API to schedule the start of an audio source on the audio thread timeline, driven by the DSP clock. https://docs.unity3d.com/ScriptReference/AudioSource.PlayScheduled.html
> 
> This allows to stitch together audio clips accurately. The use case is roughly summarized in this presentation by Jakob Schmid, audio composer/programmer of the game "140" (starting from slide 35) : https://schmid.dk/talks/2014-05-23-Nordic_Game_Conference/Schmid-2014-05-23-NGC-140.pdf
> 
> This would allow accurate audio synchronization without being restricted to specific use cases

Someone made a similar proposal in #1151

**Comment by almam72** [comment]
> > Is this proposal going to take audio that changes BPM at specific points of time into account? Or will that require a custom implementation?
> 
> Don't see why it should not work.

AudioStreams can only have one BPM set at import, no?
id also love to be able to sync to arbitrary positions seeing as you cant change BPM at runtime. i realize they're very rarely done but fadeless interactive music systems do exist and scheduling audio to arbitrary positions would make them a lot more reliable

**Comment by reduz** [comment]
@almam72 if you use the interactive player or the playlist, you can put differnt sections with different BPMs right now. Having a stream changing BPMs is definitely outside the scope, this can be made manually creating a custom stream type.

**Comment by sean-gribben** [comment]
> This might go off-topic, but if it is possible to synchronize an audio stream to the next beat or bar of another, would it be possible to synchronize a stream more arbitrarily ?
> 
> For example, Unity exposes an API to schedule the start of an audio source on the audio thread timeline, driven by the DSP clock. https://docs.unity3d.com/ScriptReference/AudioSource.PlayScheduled.html
> 
> This allows to stitch together audio clips accurately. The use case is roughly summarized in this presentation by Jakob Schmid, audio composer/programmer of the game "140" (starting from slide 35) : https://schmid.dk/talks/2014-05-23-Nordic_Game_Conference/Schmid-2014-05-23-NGC-140.pdf
> 
> This would allow accurate audio synchronization without being restricted to specific use cases

Definitely agree with what midiphony said, it should be up to the developer as to how to sync audio events, they shouldn't be restricted to a specific use case that is more complicated to implement anyway. A simple PlayScheduled would allow for any audio syncing a dev could possibly desire

**Comment by JD-The-65th** [comment]
Hey, I'm working on a Rhythm Game mapper, and to echo what @midiphony said, a PlayScheduled function that uses the DSP would be very useful to synchronize hitsounds and metronomes. A similar project to mine that's made in unity uses PlayScheduled for the aforementioned features.

Syncronizing Game Events with BPM is something that can reasonably be done (with some room for error, work still needs to be done,) and a PlayScheduled function would be a nice feature to have to accompany this proposal.
